gdnative-core/src/export/method.rs

```diff
@@ -282,18 +282,132 @@ impl<'a> Varargs<'a> {
     pub unsafe fn from_sys(num_args: libc::c_int, args: *mut *mut sys::godot_variant) -> Self {
         let args = std::slice::from_raw_parts(args, num_args as usize);
         let args = std::mem::transmute::<&[*mut sys::godot_variant], &[&Variant]>(args);
-        Self {
-            idx: 0,
-            iter: args.iter(),
-        }
+        Self { idx: 0, args }
     }
 }
 
 impl<'a> Iterator for Varargs<'a> {
     type Item = &'a Variant;
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        self.iter.next().copied()
+        if self.idx < self.args.len() {
+            let idx = self.idx;
+            self.idx += 1;
+            Some(self.args[idx])
+        } else {
+            None
+        }
+    }
+}
+
+// Return a second token.
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+
+// Count parameters.
+macro_rules! count_tts {
+    ($($tts:tt)*) => {
+        0usize $(+ replace_expr!($tts 1usize))*
+    };
+}
+
+// Convert from Varargs to tuples, implement traits.
+macro_rules! varargs_into_tuple {
+    ($($params:ident),*) => {
+        impl<'a, $($params: FromVariant),*> std::convert::TryFrom<Varargs<'a>> for ($($params,)*) {
+            type Error = VarargsError;
+
+            #[inline]
+            fn try_from(value: Varargs<'a>) -> Result<Self, Self::Error> {
+                const EXPECTED: usize = count_tts!($($params)*);
+                let args = &value.args;
+                if args.len() == EXPECTED {
+                    #[allow(unused_variables, unused_mut)]
+                    let mut i: usize = 0;
+                    #[allow(clippy::eval_order_dependence)]
+                    Ok((
+                        $({
+                            i += 1;
+                            $params::from_variant(args[i - 1])?
+                        },)*
+                    ))
+                } else {
+                    Err(ArgumentLengthError{
+                        passed: args.len(),
+                        expected: EXPECTED,
+                    }.into())
+                }
+            }
+        }
+    };
+}
+
+// Define up to the length supported by standard library.
+varargs_into_tuple!();
+varargs_into_tuple!(A);
+varargs_into_tuple!(A, B);
+varargs_into_tuple!(A, B, C);
+varargs_into_tuple!(A, B, C, D);
+varargs_into_tuple!(A, B, C, D, E);
+varargs_into_tuple!(A, B, C, D, E, F);
+varargs_into_tuple!(A, B, C, D, E, F, G);
+varargs_into_tuple!(A, B, C, D, E, F, G, H);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I, J);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I, J, K);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I, J, K, L);
```

Comment:
There's already precedent for that:

https://github.com/godot-rust/godot-rust/blob/c69c5a0241a2206ec45e2a8028922ca10648bcfa/gdnative-core/src/core_types/variant.rs#L1693

But I personally like the "explicit" version (no recursion for N-1) that you did a bit more. It's likely also more compile-time friendly, so I wouldn't change that.

You could however name the type arguments `T1..T12` (I don't know why we didn't start with 0, but I guess it's OK...).

Reply:
I could have made the code recursive, but I didn't want to make the macro too tricky.

> You could however name the type arguments T1..T12 (I don't know why we didn't start with 0, but I guess it's OK...).

The naming of my code is comes from the [standard library](https://doc.rust-lang.org/std/primitive.tuple.html). The naming of `impl_variant_for_tuples!()` probably comes from the [reference book](https://doc.rust-lang.org/reference/items/implementations.html#orphan-rules).

Well, users will not mind either naming. :upside_down_face:

Reply:
Yes, as mentioned, non-recursive is fine 🙂 

Oh, interesting theory about the origins of the namings! You're right that users won't care, but it's probably still good to be internally consistent in godot-rust. But I'm fine with leaving this for now, could be cleaned up later (as it's not a breaking change).